### PR TITLE
Making Twinkles work in the Conda DM installation

### PR DESCRIPTION
**Pull request by drphilmarshall**
This seems important, if we are going to enable casual DM users like myself to contribute to Twinkles. This PR thread is to show you my attempts.

### CODE REVIEW

**Comment by drphilmarshall** [comment]
In particular I am working on [INSTALL.md](https://github.com/DarkEnergyScienceCollaboration/Twinkles/blob/conda/INSTALL.md) and [setup.csh](https://github.com/DarkEnergyScienceCollaboration/Twinkles/blob/conda/setup/setup.csh)

**Comment by jchiang87** [comment]
Hey @drphilmarshall!  I feel your pain here.   As part of the package reorganization, we had tried to make Twinkles an eups-managed package, so in principle, getting it to work should just entail doing
```
setup Twinkles
```
but there is also some ups_db configuration that has to occur before that command will work. I made an attempt at that in the [bash version of the setup script](https://github.com/DarkEnergyScienceCollaboration/Twinkles/blob/master/setup/setup.sh), but TBH I don't know what the recommended way to do that stuff is.  In particular, the `eups declare` line is an incantation to me.  In any case, I think we should strive to get that part right.  Unfortunately, I think the DYLD_LIBRARY_PATH hack will need to be done until the conda distributions are somehow fixed.

**Comment by danielsf** [comment]
I'm not sure why you need to create ups_db or EUPS_PATH by hand.  If you are using the stack, you should already have eups setup in one form or another.

**Comment by jchiang87** [comment]
@danielsf  So how does one add Twinkles if it is a "shared" stack?  That was the problem I was trying to address with that.

**Comment by danielsf** [comment]
Off the top of my head: I'm not sure, but I have some ideas how to figure out.

What happens when you start up your shared stack and do

```
eups declare -r /path/to/Twinkles/home/directory/ Twinkles -t current
```

**Comment by danielsf** [comment]
Side note on eups declare statements:

declare is how you tell eups where a package resides.  The important commands for an eups declare statement is

-r = the path to the home directory of the package
-t = the tag (e.g. 'sims' or 'current') that you want assigned to that version of the package

So, if I did

```
eups declare -r /Users/danielsf/stack/myTwinkles/ Twinkles -t danielsf
```
I would be telling eups that a version of Twinkles resides in /Users/danielsf/stack/myTwinkles/ and that I want to tag that version as danielsf so that I can set it up via

```
setup Twinkles -t danielsf
```

The tag 'current' is special (that is what the -c in your eups declare line achieves) because 'current' is the version of a package that eups sets up if no tag is specified or if the package does not have a version corresponding to the specified tag.

eups treats 'current' with a lot of respect, so it is possible that you don't have permission to declare something 'current' by hand on a shared stack.  I'm not sure.

**Comment by jchiang87** [comment]
Here is what I see:
```
bash-4.1$ pwd -P
/nfs/slac/g/ki/ki18/jchiang/DESC/Twinkles/Twinkles
bash-4.1$ printenv EUPS_PATH
/nfs/farm/g/desc/u1/LSST_Stack_2016-02-23/lsstsw/stack
bash-4.1$ eups declare -r /nfs/slac/g/ki/ki18/jchiang/DESC/Twinkles/Twinkles Twinkles -t current
Warning: path /nfs/slac/g/ki/ki18/jchiang/DESC/Twinkles/Twinkles is absolute, not relative to EUPS_PATH
bash-4.1$ ls /nfs/farm/g/desc/u1/LSST_Stack_2016-02-23/lsstsw/stack/ups_db/Twinkles/
current.chain  tag:current.version
bash-4.1$ cat /nfs/farm/g/desc/u1/LSST_Stack_2016-02-23/lsstsw/stack/ups_db/Twinkles/tag\:current.version 
FILE = version
PRODUCT = Twinkles
VERSION = tag:current
#***************************************

Group:
   FLAVOR = Linux64
   QUALIFIERS = ""
   DECLARER = James Chiang
   DECLARED = 2016/03/04 09:52:27 PST
   PROD_DIR = /nfs/slac/g/ki/ki18/jchiang/DESC/Twinkles/Twinkles
   UPS_DIR = ups
   TABLE_FILE = Twinkles.table
End:
bash-4.1$ 
```

**Comment by danielsf** [comment]
If you are following the process documented here

https://github.com/DarkEnergyScienceCollaboration/Twinkles/blob/conda/INSTALL.md

Then you have a problem:

The ../stable/ conda channel only contains v_11 of the DM stack and a very out-dated version of the sims stack.  The sims stack conda installs live at

https://eupsforge.net/conda/dev

That would be why you are getting that failure at DBConnection (DBConnection is a relatively recent addition to the sims stack).

**Comment by danielsf** [comment]
Try
```
conda config --add channels http://eupsforge.net/conda/dev
```
and then
```
conda install lsst_sims
```
and see what happens.  You should get version 2.2.1, which contains the class DBConnection

**Comment by danielsf** [comment]
Sorry.  Those last two comments were meant for issue #164

**Comment by danielsf** [comment]
by which I mean they were meant for #160 

so.... many.... browser tabs

**Comment by danielsf** [comment]
@jchiang87 That looks like it worked.  The warning about the absolute path is harmless.  I see it all the time.

Do
```
eups list -v Twinkles
```
and see what comes out

**Comment by jchiang87** [comment]
@danielsf 
```
bash-4.1$ eups list -v Twinkles
   tag:current /nfs/farm/g/desc/u1/LSST_Stack_2016-02-23/lsstsw/stack /nfs/slac/g/ki/ki18/jchiang/DESC/Twinkles/Twinkles        current
bash-4.1$ 
```
It did work, but if someone else using that stack also does the same with their own Twinkles distribution, won't we collide?

**Comment by danielsf** [comment]
That is the next test.

Once a version of Twinkles is set up in the shared stack, it should not need to be setup again.

Log out, log back in, and see if you can setup that version of Twinkles with just the eups command

```
setup Twinkles
```

**Comment by danielsf** [comment]
Excuse me: once it is *declared*in the shared stack, it should not need to be *declared* again

(sorry; got a little excited there; things don't usually just work)

**Comment by jchiang87** [comment]
> Log out, log back in, and see if you can setup that version of Twinkles with just the eups command
> `setup Twinkles`

yes, that works.

**Comment by danielsf** [comment]
Great.  Then I think you're set, right?

FYI If you are worried about needing to declare 'current' twinkles in a different location
```
eups undeclare
```
is a command.
```
eups undeclare Twinkles -t myTag
```
will remove the version tagged 'myTag' from EUPS' version database.  This will leave you free to re-declare 'myTag' wherever you want.

**Comment by danielsf** [comment]
I guess it may be worth having a different user log in and try to setup Twinkles as above, just to make sure that the declaration wasn't limited somehow to your account.

**Comment by jchiang87** [comment]
hmm, I think we still have a problem.  We are using that shared stack for our batch processing and that processing can't use my development version (which I just declared).  How do we arrange it so that there is a default production version in that shared stack and a development version that I am using at the same time?

**Comment by danielsf** [comment]
That is the thing that EUPS is best at (allowing different versions of the same software to coexist on one system).  Just install the production version in one directory and the development version in a different directory.  You can then declare the production version tagged as 'current' and the development version tagged as, for example 'devTag'.  Then, you can setup the production version with
```
setup Twinkles
```
(remember, EUPS defaults to 'current' if no tag is specified) and setup the development version with
```
setup Twinkles -t devTag
```

EUPS is a little picky about what tags are allowed.  By default, it will only allow you to declare your own username as a tag (so you would have to tag the development version as jchiang87).  If you want to make other tags legal, you need to go into your ~.eups/ directory and add the following line to startup.py (or create a file startup.py if non exists)
```
hooks.config.Eups.userTags += ["list", "of", "tags", "I", "want", "to", "make", "legal"]
```

**Comment by jchiang87** [comment]
Thanks @danielsf !  Not understanding `eups declare` was a conceptual blocker for me to see how to have different versions of Twinkles use the same stack.   I assume we would just write protect the `Twinkles/current.chain` file to prevent people from accidently resetting the default version?

**Comment by danielsf** [comment]
That is deeper than I've ever delved into the workings of EUPS, so I'm not sure.   After poking around, though, I think it would be safest to try limiting the permissions on both
```
Twinkles/current.chain
Twinkles.current.version
```
As before, I think it would be worth it to do that, log out, log back in, then see what happens if you try to declare another version of Twinkles 'current'

**Comment by jchiang87** [comment]
I did do that logout/login in test after setting the permissions for both those files with `chmod -w`, but I was still able to run the command without error, but maybe that was because I didn't ask to change anything?  I'll try again using different location and if it fails, I'll comment here.

Thanks!

**Comment by drphilmarshall** [comment]
Success! At some cost. Check out [INSTALL.md](https://github.com/DarkEnergyScienceCollaboration/Twinkles/blob/conda/INSTALL.md) and [__init__.py](https://github.com/DarkEnergyScienceCollaboration/Twinkles/blob/conda/python/desc/twinkles/__init__.py) for a system that allows me to successfully execute the [PostageStampMaker_demo.ipynb](https://github.com/DarkEnergyScienceCollaboration/Twinkles/blob/conda/examples/notebooks/PostageStampMaker_demo.ipynb). Shall we merge this PR?

**Comment by jchiang87** [comment]
ok by me.  (consider your code reviewed! :-) )

**Comment by drphilmarshall** [comment]
W00t!
@rbiswas4 is going to add another layer of PR though, in an attempt to get
the SNe in properly.

**Comment by danielsf** [comment]
Can we talk next week about what are the barriers are to using the 
```
eups distrib install
```
 version?

While I acknowledge my opinion is the minority opinion, I am generally skeptical of the conda-installed stack.  conda-lsst (the tool used to build the conda install) is not technically an officially supported LSST product.  It is very fragile and it is unclear whose job it is to maintain it.  Mario Juric and Josh Hoblitt have been putting a lot of work into conda-lsst lately, but they both have thousands of other things they are simultaneously responsible for, so I am not optimistic that conda-lsst can keep up with the continually advancing lsst-sims stack (and I literally do not know if there is anyone who is worrying about keeping up-to-date versions of the DM stack available via conda).

Those are my concerns.  Let me know what you think.

**Comment by drphilmarshall** [comment]
I jumped on it because it seemed so much easier, and figured that many of
our potential contributors around the collaboration would find it easier to
join in by conda. I wonder how many people in DESC have conda installs
compared to eups distrib? I am certainly less keen on it than I was 24
hours ago :-)

**Comment by jchiang87** [comment]
@drphilmarshall What do you want to do about this PR?  I think the changes to INSTALL.md are useful, but in `python/desc/twinkles/__init__.py`, I think it is better to avoid importing the modules in the try/except blocks by default.  Also, as @danielsf were discussing in #188, it may be preferable to get rid of the `setup.[c]sh` scripts entirely and rely on documenting the eups usage.  In any case, now that we are using eups, the changes to `setup.csh` here are superseded.  Dealing with this would also help us to close or otherwise address #160, #165, and #189.

**Comment by drphilmarshall** [comment]
I agree that the `__init__.py` needs to work for everyone - so that alone
means that this PR should not be merged yet! I'm happy to keep my conda
branch up to date with master, but I guess the point of doing this was to
ensure that I and others could make the conda install route work. What
would you advise about this?

**Comment by danielsf** [comment]
As I see it, the principal problem with the conda install is that the DM stack and the Sims stack exist on different channels and that those channels are out of sync by at least 6 months.  Hopefully, after Michael makes his request to the project (more frequent conda builds of both Sims an DM), that will go away.  Once it does, it should be possible to draw up a Readme that teaches people enough about EUPS that they will be able to install the stack (Sims and DM), declare Twinkles to EUPS, and then setup a self-consistent Twinkles+Sims+DM stack.

I know "wait until the project answers our request" is not a great answer, but I think it is the easiest way to get a simple, usable conda-based version of Twinkles up and running.

File: INSTALL.md [hunk 1]
```diff
@@ -67,3 +78,52 @@ ImportError: dlopen(/Users/pjm/miniconda2/opt/lsst/daf_persistence/python/lsst/d
   Referenced from: /Users/pjm/miniconda2/opt/lsst/boost/lib/libboost_filesystem.dylib
   Reason: image not found
 ```
+You have this library, it's just that python doesn't know where to find it, for some reason. The solution is to add another line to your `.bashrc`, after the setup commands:
+```
+export DYLD_LIBRARY_PATH = ${LSST_DIR}/opt/lsst/boost/lib:${DYLD_LIBRARY_PATH}
```
Comment (jchiang87) [inline INSTALL.md]: The DM team suggests the following fix for this instead:

https://community.lsst.org/t/library-loading-error-in-an-ipython-notebook/623

**Comment by cwwalter** [comment]
BTW the same thing is true for using 'ipython'.  You will find you can run code where you need to import the LSST EUPS based code for the same reason.

You can 'jupyter console' instead to bring up in interactive shell that will work.

**Comment by drphilmarshall** [comment]
This branch has been overtaken by the implementation of Scott's sims channel, and various other simplifications - we now have good setup notes in the master branch at 

https://github.com/DarkEnergyScienceCollaboration/Twinkles/tree/master/doc/Setup

**Pull request closed without merging**